README.md
<p align="center"><img src="/img/socialcard.jpg" alt="Social Card of element-animation"></p>

![NPM Downloads](https://img.shields.io/npm/dt/element-animation)
![NPM](https://img.shields.io/npm/v/element-animation)
![Github Code Size](https://img.shields.io/github/languages/code-size/jersyfi/element-animation)
![GitHub License](https://img.shields.io/github/license/jersyfi/element-animation)

# element-animation

## What is this project for?
You can use this project to create congratulation confetti, snow or leaves falling from top to bottom or moved by mouse. Also elements insider your hole document can be moved by mouse movement. With this project you can create awesome effects on your website.

Start with a simple test file to see what the project can do without coding yourself. First you need to download the project and then go to the [test](./test/) folder and preview the bootstrap html file.

## Documentation

### Get started

To start with the project element-animation you need to install via NPM and then import it.

```bash
npm i element-animation
```

```javascript
import elementAnimation from element-animation

var elemAnim = new elementAnimation(
    container,
    props,
    options
)
```

container -> DOMElement (default: null) \
props -> Array (default: []) \
options -> Object (default: {})

Before we start to define the variables you need to know that you only need `container` and `props` when using the functions `rain()`, `mousemove()` and `rainWithMousemove()`.
First we start with the container. Here you need to give a DOMElement for example a simple `<div></div>` container.

```html
<div id="element-animation"></div>
```

```javascript
const container = document.getElementById('element-animation')
```

Then you need to define props. There you give an array with the sources to the images like .jpg or .svg.

```javascript
props = [
    'src_to_img',
    'src_to_img_1'
]
```

For the options you have the following possibilities. When a comment is behind a option it means that the option is only for a specific funtion.

```javascript
options = {
    interval: 550,          // rain
    rotate: [-40, 40],
    size: [10, 15],
    duration: [5, 13],      // rain
    amount: [30, 40]        // mousemove
}
```

Now you need to call one of the given functions [described down below](#functions) to visially see the animation. Choose between `rain()`, `mousemove()`, `rainWithMousemove()` and a special function `mousemoveElements()`.

### Functions

All functions except `mousemoveElements()` need a container for the generated elements.

#### rain()
```javascript
new elementAnimation(container, props, options).rain()
```

#### mousemove()
```javascript
new elementAnimation(container, props, options).mousemove()
```

#### rainWithMousemove()
```javascript
new elementAnimation(container, props, options).rainWithMousemove()
```

#### mousemoveElements()
data-p-element -> Defines that it is a element for movement
<br>data-p-move -> Factor from 0 to unlimited (float with dots allowed)

```html
<div data-ea-element data-ea-move="0.5">
    Example text or element here.
</div>

<div data-ea-element data-ea-move="0.8">
    Secound example text or element here.
</div>
```

```javascript
new elementAnimation(container, props, options).mousemoveElements()
```

### Helpers

#### reset()
Can be used to reset all prop actions.
After a reset you can call another functions

```javascript
const elemAnim = new elementAnimation(container, props, options).rain()

elemAnim.reset()
elemAnim.mousemove()
```

## Changelog
Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Credits
- [Jérôme Bastian Winkel](https://github.com/jersyfi)
- [All Contributors](../../contributors)

## License
The MIT License (MIT). Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to `element-animation` will be documented in this file.

## v1.0.1 - 2021-07-15

- Fixed wrong query name

## v1.0.0 - 2021-07-14

- Fixed bug with function `rainWithMoousemove()`
- Modified Documentation

## v0.2.2 - 2021-07-07

- Modified Documentation

## v0.2.1 - 2021-07-06

- Modified Documentation
- Modified package.json

## v0.2.0 - 2021-07-06

- Renamed project from `propsjs` to `element-animation`
- Added special function `mousemoveElements()`
- Modified Documentation
- Modified Test file for new function
- Bug fix with element rotation

## v0.0.1 - 2021-06-25

- Initial uploading


test/index.js
import elementAnimation from '../src/index.js'

const container = document.getElementById('element-animation'),
    props = [
        './props/confetti_large_big_blue.svg',
        './props/confetti_large_green.svg',
        './props/confetti_small_big_yellow.svg',
        './props/confetti_large_big_red.svg',
        './props/confetti_large_blue.svg',
        './props/confetti_small_red.svg'
    ],
    options = {
        interval: 350,      // rain
        rotate: [-40, 40],
        size: [10, 15],
        duration: [2, 8],   // rain
        amount: [30, 40]    // mousemove
    },
    radioRain = document.getElementById('rain'),
    radioMousemove = document.getElementById('mousemove'),
    radioRainWithMousemove = document.getElementById('rainWithMousemove'),
    radioMousemoveElements = document.getElementById('mousemoveElements'),
    elemAnimType = {
        0: 'rain',
        1: 'mousemove',
        2: 'rainWithMousemove',
        3: 'mousemoveElements'
    },
    elemAnimState = localStorage.getItem('element-animation')

var elemAnim = new elementAnimation(container, props, options)

let eventRain = () => {
    radioRain.checked = true
    localStorage.setItem('element-animation', elemAnimType[0])
    elemAnim.reset()
    elemAnim.rain()
}

let eventMousemove = () => {
    radioMousemove.checked = true
    localStorage.setItem('element-animation', elemAnimType[1])
    elemAnim.reset()
    elemAnim.mousemove()
}

let eventRainWithMousemove = () => {
    radioRainWithMousemove.checked = true
    localStorage.setItem('element-animation', elemAnimType[2])
    elemAnim.reset()
    elemAnim.rainWithMousemove()
}

let eventMousemoveElements = () => {
    radioMousemoveElements.checked = true
    localStorage.setItem('element-animation', elemAnimType[3])
    elemAnim.reset()
    elemAnim.mousemoveElements()
}

radioRain.addEventListener('change', eventRain)
radioMousemove.addEventListener('change', eventMousemove)
radioRainWithMousemove.addEventListener('change', eventRainWithMousemove)
radioMousemoveElements.addEventListener('change', eventMousemoveElements)

if (elemAnimState == elemAnimType[1]) eventMousemove()
else if (elemAnimState == elemAnimType[2]) eventRainWithMousemove()
else if (elemAnimState == elemAnimType[3]) eventMousemoveElements()
else eventRain()


src/index.js
export default class elementAnimation {
    constructor(container = null, props = [], options = {}) {
        this.container = container
        this.props = props
        this.options = options

        this.container.style.height = '100vh'
        this.container.style.width = '100vw'
        this.container.style.position = 'fixed'
        this.container.style.overflow = 'hidden'
        this.container.style.userSelect = 'none'
        this.container.style.pointerEvents = 'none'
    }

    /**
     * Prop rain from top to bottom
     */
    rain() {
        let render = () => {
            const Elem = document.createElement('div'),
                img = document.createElement('img'),
                selectedProp = this.props[Math.floor(Math.random() * this.props.length)]

            var r = this.options.rotate,
                d = this.options.duration,
                s = this.options.size

            var rotate = this.getRandomInt(r ? r[0] : -40, r ? r[1] : 40),
                duration = this.getRandomInt(d ? d[0] : 5, d ? d[1] : 13)

            Elem.classList.add('element-animation')
            Elem.style.left = Math.random() * 100 + '%'
            Elem.style.position = 'fixed'
            Elem.style.zIndex = 45
            Elem.style.top = '-20vh'
            Elem.animate([
                { transform: 'translateY(0)' },
                { transform: 'translateX(' + this.getRandomInt(5, 5)  + 'vw) translateY(120vh)' }
            ], {
                duration: duration * 1000,
            })
            img.animate([
                { transform: 'rotate(' + rotate  + 'deg)' },
                { transform: 'rotate(' + this.getRandomInt(rotate - 20, rotate + 20)  + 'deg)' }
            ], {
                duration: duration * 1000,
            })
            img.width = this.getRandomInt(s ? s[0] : 10, s = s ? s[1] : 15)
            img.src = selectedProp
            Elem.appendChild(img)
            this.container.appendChild(Elem)

            setTimeout(() => {
                Elem.remove()
            }, 15000)
        }

        window.elementAnimationInterval = setInterval(render, this.options.interval ?? 550)
    }

    /**
     * Props mouving by mouse movement
     */
    mousemove() {
        let render = () => {
            var a = this.options.amount,
                r = this.options.rotate,
                s = this.options.size

            window.propAmount = this.getRandomInt(a ? a[0] : 30, a ? a[1] : 40)

            for (let i = 0; i < window.propAmount; i++) {
                const Elem = document.createElement('div'),
                    img = document.createElement('img'),
                    selectedProp = this.props[Math.floor(Math.random() * this.props.length)]

                Elem.classList.add('element-animation')
                Elem.classList.add('element-animation-mousemove-' + i)
                Elem.style.position = 'fixed'
                Elem.style.zIndex = 45
                Elem.style.top = Math.random() * 100 + 'vh'
                Elem.style.left = Math.random() * 100 + 'vw'
                Elem.setAttribute('data-ea-move', Math.random() * 10)
                img.style.transform = 'rotate(' + this.getRandomInt(r ? r[0] : -40, r ? r[1] : 40) + 'deg)'
                img.width = this.getRandomInt(s ? s[0] : 10, s = s ? s[1] : 15)
                img.src = selectedProp

                Elem.appendChild(img)
                this.container.appendChild(Elem)
            }

            document.body.addEventListener('mousemove', this.handleMousemove)
        }

        render()
    }

    /**
     * Prop rain from top to bottom and moving by mouse movement
     */
    rainWithMousemove() {
        let render = () => {
            const Elem = document.createElement('div'),
                img = document.createElement('img'),
                selectedProp = this.props[Math.floor(Math.random() * this.props.length)]

            var r = this.options.rotate,
                d = this.options.duration,
                s = this.options.size

            var rotate = this.getRandomInt(r ? r[0] : -40, r ? r[1] : 40),
                duration = this.getRandomInt(d ? d[0] : 5, d ? d[1] : 13)

            Elem.classList.add('element-animation')
            Elem.classList.add('element-animation-rainwithmousemove')
            Elem.style.left = Math.random() * 100 + '%'
            Elem.style.position = 'fixed'
            Elem.style.zIndex = 45
            Elem.style.top = '-20vh'
            Elem.animate([
                { transform: 'translateY(0)' },
                { transform: 'translateX(' + this.getRandomInt(5, 5)  + 'vw) translateY(120vh)' }
            ], {
                duration: duration * 1000,
            })
            Elem.setAttribute('data-ea-move', Math.random() * 10)
            img.animate([
                { transform: 'rotate(' + rotate  + 'deg)' },
                { transform: 'rotate(' + this.getRandomInt(rotate - 10, rotate + 10)  + 'deg)' }
            ], {
                duration: duration * 1000,
            })
            img.width = this.getRandomInt(s ? s[0] : 10, s = s ? s[1] : 15)
            img.src = selectedProp
            Elem.appendChild(img)
            this.container.appendChild(Elem)

            setTimeout(() => {
                Elem.remove()
            }, 15000)

            document.body.addEventListener('mousemove', this.handleRainWithMousemove)
        }

        window.elementAnimationInterval = setInterval(render, this.options.interval ?? 550)
    }

    mousemoveElements() {
        document.body.addEventListener('mousemove', this.handleMousemoveElements)
    }

    /**
     * Helper functions
     */

    /**
     * Return random int value with given min and max
     * 
     * @param {*} min 
     * @param {*} max 
     * @returns
     */
    getRandomInt(min, max) {
        min = Math.ceil(min)
        max = Math.floor(max)
        
        return Math.floor(Math.random() * (max - min)) + min
    }

    /**
     * Resetting the props
     */
    reset() {
        this.container.innerHTML = ''
        clearInterval(window.elementAnimationInterval)
        document.body.removeEventListener('mousemove', this.handleMousemove)
        document.body.removeEventListener('mousemove', this.handleRainWithMousemove)
        document.body.removeEventListener('mousemove', this.handleMousemoveElements)
    }

    /**
     * Event handler
     */

    /**
     * Handles the mousemove event
     * 
     * @param {*} event 
     */
    handleMousemove(event) {
        const ww = window.innerWidth,
            wh = window.innerHeight,
            cx = event.clientX,
            cy = event.clientY

        var calc = (inner, client) => {
            return (client - (inner / 2)) / (inner * 0.05)
        }

        for (let i = 0; i < window.propAmount; i++) {
            const Elem = document.getElementsByClassName('element-animation-mousemove-' + i)[0],
                x = calc(ww, cx) * (Elem.getAttribute('data-ea-move') / 5.5),
                y = calc(wh, cy) * (Elem.getAttribute('data-ea-move') / 5.5)

            Elem.style.transform = 'translateX(' + x + 'vw) translateY(' + y + 'vh)'
        }
    }

    /**
     * Handles the rainWithMousemove event
     * 
     * @param {*} event 
     * @param {*} Elem 
     */
    handleRainWithMousemove(event) {
        const ww = window.innerWidth,
            wh = window.innerHeight,
            cx = event.clientX,
            cy = event.clientY

        var calc = (inner, client) => {
            return (client - (inner / 2)) / (inner * 0.05)
        }

        const Elems = document.getElementsByClassName('element-animation-rainwithmousemove')

        for (const Elem of Elems) {
            const x = calc(ww, cx) * (Elem.getAttribute('data-ea-move') / 5.5),
                y = calc(wh, cy) * (Elem.getAttribute('data-ea-move') / 5.5)

            Elem.style.marginLeft = x + 'px'
            Elem.style.marginTop = y + 'px'
        }        
    }

    handleMousemoveElements(event) {
        const ww = window.innerWidth,
            wh = window.innerHeight,
            cx = event.clientX,
            cy = event.clientY

        const Elems = document.querySelectorAll('[data-ea-element]')

        var calc = (inner, client) => {
            return (client - (inner / 2)) / (inner * 0.05)
        }

        for (const Elem of Elems) {
            const move = Elem.getAttribute('data-ea-move') ?? 1

            const x = calc(ww, cx) * (move / 5.5),
                y = calc(wh, cy) * (move / 5.5)

            Elem.style.transform = 'translateX(' + x + 'vw) translateY(' + y + 'vh)'
        }
    }
}
